tests/unit/common/eeprom/eeprom_acces.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+#include <vector>
+#include <stdint.h>
+#include "crc32.h"
+#include <stdlib.h>
+#include <string.h>
+#include "msgpack.h"
+#include <optional>
+
+// helper structs
+struct Key {
+    uint32_t key;
+    template <class Packer>
+    void pack(Packer &pack) {
+        pack(key);
+    }
+};
+
+template <class T>
+struct DataItem {
+    uint32_t key;
+    T data;
+
+    DataItem(uint32_t key, T data)
+        : key(key)
+        , data(data) {}
+    DataItem() = default;
+
+    template <class Packer>
+    void pack(Packer &pack) {
+        pack(key, data);
+    }
+};
+
+class EepromAccess {
+#ifdef EEPROM_UNITTEST
+public:
+#endif
+    static constexpr uint16_t START_OFFSET = 0x500;
+    static constexpr uint8_t CUSTOM_ITEM_ID = 0x7f;
+    static constexpr uint8_t MSGPACK_EXT8 = 0xc7;
+    static constexpr uint8_t LAST_ITEM_STOP = 0xfe;
+    static constexpr size_t HEADER_SIZE = 3;
+    static constexpr size_t CRC_SIZE = 4;
+
+    std::vector<std::pair<uint32_t, uint16_t>> index;
+
+    uint16_t first_free_space = START_OFFSET;
+    uint16_t write_stop = 8096; //<FIXME should be set as part of init proccess
+
+    template <class T>
+    std::vector<uint8_t> static serialize_data(const char *key, T data);
+    template <class T>
+    DataItem<T> static deserialize_data(std::vector<uint8_t> data);
+
+    // find address for hash in index
+    std::optional<uint16_t> get_address(uint32_t hash);
+
+    // stores item to eeprom and calculates the crc of the data
+    // creates header of the data
+    // the header  element ext 8 from msgpack spec and the value for type is EepromAccess::CUSTOM_ITEM_ID
+    // +--------+--------+--------+========+
+    // |  0xc7  |XXXXXXXX|  type  |  data  |
+    // +--------+--------+--------+========+
+    // where XXXXXXXX is a 8-bit unsigned integer which represents the length of the data including crc
```

Comment:
What's the 0xc7 thing? And, is it possible to combine the type and this tag together? :thinking: I mean, the eeprom is not that big, so maybe being able not to waste the space would be nice.

Reply:
This is ext 8 type from msgpack spec, I could remove the type and just use the 0xc7, but then we will not be compatible with msgpack.

Reply:
Well, we can get _inspired_ by msgpack, or reuse parts of it. If adhering to it means having some 25% ineffeciency ratio, then it's a bit expensive.

So we can have length-prefixed chunk (very simple encoding) and msgpack inside, it would still give us most of the benefits.

Reply:
I aggree, I willl change it to just lenght prefix.